### DISCUSSION TITLE
Saturation constraint

### DISCUSSION THREAD

**Discussion by halakoom** [do-mpc/do-mpc#388] [Q&A]
Hi 
I have the following states:
```python
n11 = model.set_variable(var_type = '_x',var_name='n11', shape=(1,1))
n12 = model.set_variable(var_type = '_x',var_name='n12', shape=(1,1))
n13 = model.set_variable(var_type = '_x',var_name='n13', shape=(1,1))
n14 = model.set_variable(var_type = '_x',var_name='n14', shape=(1,1))
```

and the following expression, which is simply the sum of all states
```python
n1 = model.set_expression('n1',model.x['n11']+model.x['n12']+model.x['n13']+model.x['n14'])
```

I added the following upper bound for the expression
```python
 mpc.set_nl_cons('n1cap',model.aux['n1'],ub=16e3,soft_constraint=False)
```

However, when I run the simulation, I see n1 (the expression) is larger than the upper bound I set here.

**Comment by 4flixt** [comment]
Hi @halakoom ,

can you confirm that ipopt found an optimal solution? Otherwise any behaviour is possible. 
Also what do you mean with "run the simulation"? Do you mean the closed-loop behavior including MPC + Simulator?
Because the simulator by itself does not consider this constraint. 

Finally, as a hint, you make your notation more concise with:
```python
n1 = model.set_variable(var_type = '_x',var_name='n1', shape=(4,1))
```
and have
```python
import casadi as cas

n1_sum = model.set_expression('n1', cas.sum1(n1))
```
Best,
4flixt

**Reply by halakoom** [reply]
Hi, the ipopt says it found a solution. However, the n1_sum constraint is violated. I did this in a closed-loop simulation (mpc+simulator). Is there any way to make n1 hit a saturation value like: n1 = min(n1,njam). Meaning if n1 gets larger than njam, it will be replaced by njam. I tried putting it as a constraint, but since n1 is an algebraic variable, I couldn't do any min/max function on it.
